File: gitdot-web/packages/api/src/endpoint/review/review-review-diff.ts
import { z } from "zod";
import { ReviewResource } from "../../resource";
import type { Endpoint } from "../endpoint";
import { ReviewCommentInput } from "./create-review-comments";

export { ReviewCommentInput };

export const ReviewReviewDiffRequest = z.object({
  action: z.enum(["comment", "approve", "reject"]),
  comments: z.array(ReviewCommentInput),
});
export type ReviewReviewDiffRequest = z.infer<typeof ReviewReviewDiffRequest>;

export const ReviewReviewDiffResponse = ReviewResource;
export type ReviewReviewDiffResponse = z.infer<typeof ReviewReviewDiffResponse>;

export const ReviewReviewDiff = {
  path: "/repository/{owner}/{repo}/review/{number}/diff/{position}/review",
  method: "POST",
  request: ReviewReviewDiffRequest,
  response: ReviewReviewDiffResponse,
} as const satisfies Endpoint;
export type ReviewReviewDiff = typeof ReviewReviewDiff;

File: gitdot-api/src/endpoint/review/review_review_diff.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{endpoint::Endpoint, resource::review::ReviewResource};

pub struct ReviewReviewDiff;

impl Endpoint for ReviewReviewDiff {
    const PATH: &'static str = "/repository/{owner}/{repo}/review/{number}/diff/{position}/review";
    const METHOD: http::Method = http::Method::POST;

    type Request = ReviewReviewDiffRequest;
    type Response = ReviewReviewDiffResponse;
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReviewReviewDiffCommentInput {
    pub revision_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub body: String,
    pub file_path: Option<String>,
    pub line_number_start: Option<i32>,
    pub line_number_end: Option<i32>,
    pub start_character: Option<i32>,
    pub end_character: Option<i32>,
    pub side: Option<String>,
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct ReviewReviewDiffRequest {
    pub action: String,
    pub comments: Vec<ReviewReviewDiffCommentInput>,
}

pub type ReviewReviewDiffResponse = ReviewResource;
